Issue 54628

Shubhangi2901 opened the pull request:
PR for Enable geo_distance and geo_bounding_box queries on geo_shape field type

- I have added validation in geoDistanceQueryBuilder and geoBoundingBoxQueryBuilder

Shubhangi2901 committed:
adding abstract geofieldtype in bounding box queries

Shubhangi2901 committed:
add validation queries for GeoBoundingBox Query Builder

Shubhangi2901 committed:
add validation of abstractgeometryfieldtype in geo distance

Shubhangi2901 committed:
Update GeoBoundingBox.java

Shubhangi2901 committed:
Update GeoBoundingBox.java

Shubhangi2901 committed:
Update StreamInput.java

iverase commented on server/src/main/java/org/elasticsearch/index/query/GeoBoundingBoxQueryBuilder.java:
```diff
@@ -312,10 +327,15 @@ public Query doToQuery(QueryShardContext context) {
         }
-        if (!(fieldType instanceof GeoPointFieldType)) {
-            throw new QueryShardException(context, "field [" + fieldName + "] is not a geo_point field");
+//        if (!(fieldType instanceof GeoPointFieldType)) {
```
Note that above refers just to geo_point ("failed to find geo_point field ["). That might need to be updated.

iverase commented:
Validation looks good. Next thing is to change the way we are building the queries by using `AbstractGeometryFieldType#geometryQueryBuilder()`

Shubhangi2901 committed:
geodistance and geoboundingbox queries now inherit from AbstractGeometryQueryBuilder

Shubhangi2901 committed:
Merge branch 'Issue-54628' of https://github.com/Shubhangi2901/elasticsearch into Issue-54628

Shubhangi2901 commented:
Hi @iverase,

I have updated GeoBoundingBox and GeoDistance to inherit and implement AbstractGeometryQueryBuilder as given in GeoShapeQueryBuilder. I have added the following functions from GeoShapeQuery:
1. buildShapeQuery(QueryShardContext context, MappedFieldType fieldType)
2. GeoDistanceQueryBuilder fromXContent(XContentParser parser)

Let me know if I have progressed in the right direction. Also if new issues should be created to update docs with new changes.

iverase commented:
Thanks @Shubhangi2901!

I had a look into your changes and I don't think this queries should implement `AbstractGeometryQueryBuilder` as these are very specific queries and they do not do as much things as that abstraction. I think it is enough to just update the way we are building the queries, so for example for geo_circle queries will be enough to update the query logic to:

```
       final AbstractGeometryFieldMapper.AbstractGeometryFieldType ft =
            (AbstractGeometryFieldMapper.AbstractGeometryFieldType) fieldType;

        Circle circle = new Circle(center.lon(), center.lat(), this.distance);
        return ft.geometryQueryBuilder().process(circle, fieldType.name(), SpatialStrategy.RECURSIVE,
            ShapeRelation.INTERSECTS, shardContext);
``` 

Note that we need to update the test so all the possibilities are tested.

elasticmachine commented:
Pinging @elastic/es-analytics-geo (:Analytics/Geo)

iverase commented:
I am going to close this PR in favour of #64224.
